src/commands/mod.rs

```diff
@@ -1,4 +1,34 @@
+use std::fmt::{Display, Formatter};
+use std::io;
+use std::path::PathBuf;
+
 pub(crate) mod generate_buildpack_matrix;
 pub(crate) mod generate_changelog;
 pub(crate) mod prepare_release;
 pub(crate) mod update_builder;
+
+pub(crate) fn resolve_path(value: Option<PathBuf>) -> Result<PathBuf, ResolvePathError> {
```

Comment:
When a function only takes an optional argument I immediately suspect something is off with the design since there are only very few use-cases where that signature makes logical sense. In many cases it's hiding an implicit default deep in the code.

This is the case here too. Since the `working_dir` field of the user-provided args is optional it must've been more convenient to design `resolve_path` so that it can just take that value verbatim. However, this moves the _business logic_ of "when no working dir is specified, the current dir is used" down into a utility function - entangling these concerns across the call stack.

In addition, there are call-sites where the `value` passed in is not an `Option`. This results in the need to add `Some()` at the call site which serves no logical purpose in that context.

We should move the implicit default up where it belongs - into `GenerateBuildpackMatrixArgs`. Adding a function to `GenerateBuildpackMatrixArgs` that does the resolution of the path would go a long way. I imagine the signature to look something like this: `fn working_dir(&self) -> std::io::Result<PathBuf>`.

With the default value concern out of the way, we can refactor `resolve_path`. First, we can rename it to something like `absolute_path` and change the signature to something like `fn absolute_path(path: &Path, base: &Path) -> PathBuf`. This would make the function pure (with the immediate benefit that is much easier to test).

Where did the error go? Well, the only case where this can fail is if we need to determine current directory. We moved this out of this function to where we actually need to know the current directory - when resolving the working directory specified by the user. This means that at call-sites where we always passe `Some(foo)` the need to handle an error that could actually never logically occur goes away entirely!

I think this is an excellent case study of the impact of designing functions so that they make sense in a vacuum. This has positive ripple effects throughout the code that go way beyond the function itself. In the case here:

- Readers looking at `GenerateBuildpackMatrixArgs` can now immediately see what the default value for `working_dir` is.
- `resolve_path`/`absolute_path` is now pure and easily testable because there are no side-effects.
- Users of `resolve_path`/`absolute_path` don't need to deal with errors when no error can actually occur.
- We can remove the `ResolvePathError` struct.

Reply:
This one came out of the desire to have several of the command args (`GenerateBuildpackMatrixArgs`, `GenerateChangelogArgs`, and `PrepareReleaseArgs`) allow for the option to change their working directory or default to the current one. With no easy way to handle this as a default in `clap` the function started out with the `Option<PathBuf>` for convenience. Then later, when I needed almost the exact same functionality with a non-optional path I just repurposed the function instead of refining it.

I've broken down the single `resolve_path` function which is currently doing too much into the following:
- a pure `resolve_path(path: &Path, base: &Path) -> PathBuf` function that, as you say, makes sense in a vacuum and does not error
- a convenience `resolve_working_dir_from_current_dir(value: Option<PathBuf>) -> std::io::Result<PathBuf>` function that handles resolving the optional `working_dir` argument for the various `clap` commands

I've also dropped the `ResolvePathError`.

Reply:
I'm not sure I understand. If we'd add functions to get either the explicitly configured working directory value or the default (`std::env::current_dir`) to `GenerateBuildpackMatrixArgs`, `GenerateChangelogArgs`, and `PrepareReleaseArgs`, we can work with `resolve_path` in all cases and don't need a highly specific helper (plus the need to come up with a name for it) that does hide the default far away from where it should be logically placed.

Reply:
I think you're misunderstanding the intent. But, as it stands, I don't want to hold up this PR for a minor detail so I've removed the function.